### PR TITLE
Refactor TypeInfo into a variant in luajit-ffi-gen

### PR DESCRIPTION
**Pull request by dgavedissian**
This makes it far more explicit when we handle plain types vs options vs boxes etc. I think this structure also makes sense as not all types (box, functions, etc) support refs.

I used the opportunity to massively increase our test coverage for luajit-ffi-gen. My goal was to attempt to cover almost all cases in the tests.

I also added the ability to configure the location of the Lua FFI code generation, which allowed me to actually generate and commit the Lua FFI code for the unit test structs. Obviously, this code is not used in the game itself, but it does mean that we can see the ffi_gen and ffi_meta files for all possible cases without relying on the game code itself.

### CODE REVIEW

File: engine/lib/luajit-ffi-gen/README.md [hunk 1]
```diff
@@ -341,10 +341,9 @@ List of allowed types in the input parameter position.
 
 | Rust type                    | extern "C" interface       | C type            |
 | ---------------------------- | -------------------------- | ----------------- |
-| By value (NT) (not working!) | Box\<NT>                   | NT*               |
+| By value (MT) (not working!) | Box\<MT>                   | MT*               |
```
Comment (dgavedissian) [inline engine/lib/luajit-ffi-gen/README.md:344]: In a future PR, I will make this work using the ideas discussed on Discord.

**Review by IllustrisJack — APPROVED** [review]
Seeing no apparent problems with this

File: engine/lib/luajit-ffi-gen/src/impl_item/generate_rust.rs [hunk 2]
```diff
@@ -90,15 +90,15 @@ impl ImplInfo {
     // Note: We return a list of token streams here, because a single parameter can generate multiple parameters in the wrapper function.
     fn gen_wrapper_param(&self, param: &ParamInfo) -> Vec<TokenStream> {
         let name_ident = format_ident!("{}", param.name);
-        let param_type = param.ty.as_ffi(&self.name).rust;
+        let param_type = param.ty.as_ffi(&self.name).0;
         let param_type_tokens: TokenStream =
             param_type.parse().expect("Unable to parse Rust FFI type");
```
Comment (HaronK) [inline engine/lib/luajit-ffi-gen/src/impl_item/generate_rust.rs:95]: I think that was me who added expect and panic in the generators but there should be more correct way to do it. Like in the parsing part we return an error. So probably for generation part there is a similar way.
But that's for later.

File: engine/lib/phx/src/event_bus/payload.rs [hunk 3]
```diff
@@ -311,10 +311,9 @@ impl EventPayload {
         value.iter().for_each(|v| f(*v));
     }
 
-    // TODO: uncomment when luajit_gen_ffi supports array/slices of strings properly
```
Comment (HaronK) [inline engine/lib/phx/src/event_bus/payload.rs:314]: Uncomment also a **TODO** part in `EventPayloadConverter:valueToPayload`. It relates to this one.

**Review by HaronK — APPROVED** [review]
Overall looks good. Can be merged after addressing several comments.

In total it's a great job @dgavedissian. Especially amount of tests.
When I initially implemented this, I was too tired (and lazy :P ) to add more test, and I relied in general on actual usage. So it's good to have better test coverage.
I think for now it should be enough tests. We will add more when this macro will be used more widely.

**Comment by dgavedissian** [comment]
> When I initially implemented this, I was too tired (and lazy :P ) to add more test

I completely understand :D it was a struggle to write tests here too, but I was afraid to accidentally break something without adding them first.

**Pull request merged**